tree.go

```diff
@@ -1075,15 +1097,76 @@ func (n *LeafNode) updateMultipleLeaves(values [][]byte) {
 	}
 }
 
-func (n *LeafNode) Delete(k []byte, _ NodeResolverFn) error {
+// Delete deletes a value from the leaf, return `true` as a second
+// return value, if the parent should entirely delete the child.
+func (n *LeafNode) Delete(k []byte, _ NodeResolverFn) (error, bool) {
 	// Sanity check: ensure the key header is the same:
 	if !equalPaths(k, n.stem) {
-		return nil
+		return nil, false
 	}
 
-	var zero [32]byte
-	n.updateLeaf(k[31], zero[:])
-	return nil
+	// Erase the value it used to contain
+	original := n.values[k[31]] // save original value
+	n.values[k[31]] = nil
+
+	// Check if a Cn subtree is entirely empty, or if
+	// the entire subtree is empty.
+	var (
+		isCnempty = true
+		isCempty  = true
+	)
+	for i := 0; i < NodeWidth; i++ {
+		if len(n.values[i]) > 0 {
+			// if i and k[31] are in the same subtree,
+			// set both values and return.
+			if byte(i/128) == k[31]/128 {
+				isCnempty = false
+				isCempty = false
+				break
+			} else {
+				// i and k[31] were in a different subtree,
+				// so all we can say at this stage, is that
+				// the whole tree isn't empty.
+				// TODO if i < 128, then k[31] >= 128 and
+				// we could skip to 128, but that's an
+				// optimization for later.
```

Comment:
Yes, or be a bit smarter in L1118 (for initialization) to only iterate in the right subtree.
e.g: `i:=k[31]/12*128; i<k[31]/12*128+128; i++` ish
Prob requires some extra `C1 vs C2 != empty` checks to fully complete the optimization.

Reply:
Ah yes good point, I could indeed check that `Cn != zero`, and then only go over one subtree :+1:

Reply:
not going to fix that today though